File: build_docs.py
import os
import re
import shutil

# --- Configuration ---
SOURCE_FILE = 'ver.2.0.1.0/ver.2.0.1.0.txt'
DOCS_DIR = 'medical-wiki/docs'
CATEGORY_PATTERN = r'={64}\n【(No\.\d+)】(.*?)\n={64}'
DISEASE_PATTERN = r'-{32}'

def slugify(text):
    """Simple kebab-case converter for directory/filenames."""
    text = text.lower()
    text = re.sub(r'[^a-z0-9]+', '-', text)
    return text.strip('-')

def parse_medical_data():
    if not os.path.exists(SOURCE_FILE):
        print(f"Error: {SOURCE_FILE} not found.")
        return

    with open(SOURCE_FILE, 'r', encoding='utf-8') as f:
        content = f.read()

    # Split into categories
    # Finding matches for Categories
    categories = re.split(CATEGORY_PATTERN, content)
    # re.split with groups returns [prefix, group1, group2, content_after, ...]
    
    # Clean up docs dir (optional, but requested in spec Step 4)
    # We'll just append for now and let user decide if they want a wipe.
    # Spec says "削除せよ" for default docs, so we'll do that later.

    # Skip first element if it's junk
    idx = 1
    while idx < len(categories):
        cat_id = categories[idx]
        cat_name = categories[idx+1]
        cat_content = categories[idx+2]
        idx += 3

        cat_dir_name = f"{cat_id.replace('No.', '')}-{slugify(cat_name)}"
        cat_path = os.path.join(DOCS_DIR, cat_dir_name)
        os.makedirs(cat_path, exist_ok=True)

        # Split category content into diseases
        diseases = re.split(DISEASE_PATTERN, cat_content)
        
        disease_count = 0
        for disease_block in diseases:
            disease_block = disease_block.strip()
            if not disease_block:
                continue

            # Parse Disease Info
            # Look for [ICD] Name
            title_match = re.search(r'\[(.*?)\] (.*?)\n', disease_block)
            if not title_match:
                continue
            
            icd_code = title_match.group(1)
            disease_name = title_match.group(2)
            disease_count += 1

            # Extract sections
            s_text = extract_section(disease_block, '<S:', ['<O:', '<Ns:', '<A:', '<P:', '▶ Dr.Advice'])
            o_text = extract_section(disease_block, '<O:', ['<S:', '<Ns:', '<A:', '<P:', '▶ Dr.Advice'])
            ns_text = extract_section(disease_block, '<Ns:', ['<S:', '<O:', '<A:', '<P:', '▶ Dr.Advice'])
            a_text = extract_section(disease_block, '<A:', ['<S:', '<O:', '<Ns:', '<P:', '▶ Dr.Advice'])
            p_text = extract_section(disease_block, '<P:', ['<S:', '<O:', '<Ns:', '<A:', '▶ Dr.Advice'])
            
            # Dr.Advice (multiple items possible)
            advices = re.findall(r'▶ Dr\.Advice (.*?)\n', disease_block)
            advice_html = ""
            if advices:
                advice_html = ":::tip Dr.Advice\n" + "\n".join([f"- {a.strip()}" for a in advices]) + "\n:::"

            # Prepare Filename
            safe_name = slugify(f"{icd_code}-{disease_name}")
            file_path = os.path.join(cat_path, f"{safe_name}.mdx")

            # Generate MDX Content
            mdx_content = f"""---
id: {safe_name}
title: "[{icd_code}] {disease_name}"
sidebar_position: {disease_count}
---

<div className="row">
  <div className="col col--7">
    ## 📝 診療テンプレート
    
    ### <S: 主訴・ROS>
    {s_text.strip()}

    ### <O: 所見・検査>
    {o_text.strip()}

    ### <Ns: 診察前対応>
    {ns_text.strip()}

    ### <A: 評価>
    {a_text.strip()}

    ### <P: 方針>
    {p_text.strip()}

    {advice_html}
  </div>
  
  <div className="col col--5">
    ## 📖 教科書的詳細解説
    :::info Wiki編集エリア
    この右側のエリアは、最新のガイドラインや病態生理、エビデンスを追記するためのスペースです。
    ページ下部の「このページを編集」から追記・修正を行ってください。
    :::
    
    ### 🔗 関連リンク・文献
    - [国内ガイドライン検索](https://www.google.com/search?q={disease_name}+ガイドライン)
  </div>
</div>
"""
            with open(file_path, 'w', encoding='utf-8') as f_out:
                f_out.write(mdx_content)

    print("Parsing completed.")

def extract_section(text, start_marker, next_markers):
    start_idx = text.find(start_marker)
    if start_idx == -1:
        return ""
    
    # Move start_idx to after the marker
    # Wait, the spec shows markers like <S: 主訴・ROS>
    # The actual text might have trailing colon or space
    content_start = text.find('\n', start_idx)
    if content_start == -1:
        content_start = start_idx + len(start_marker)
    
    # Find the nearest next marker
    end_idx = len(text)
    for marker in next_markers:
        pos = text.find(marker, content_start)
        if pos != -1 and pos < end_idx:
            end_idx = pos
            
    return text[content_start:end_idx].strip()

if __name__ == "__main__":
    parse_medical_data()

File: build_docs.mjs
import fs from 'fs';
import path from 'path';

// --- Configuration ---
const SOURCE_FILE = 'ver.2.0.1.0/ver.2.0.1.0.txt';
const DOCS_DIR = 'medical-wiki/docs';
const CATEGORY_PATTERN = /={64}\n【(No\.\d+)】(.*?)\n={64}/g;
const DISEASE_PATTERN = /-{32}/;

function slugify(text) {
    if (!text) return "";
    return text.toLowerCase()
        .replace(/[^a-z0-9]+/g, '-')
        .replace(/^-+|-+$/g, '');
}

function escapeJS(text) {
    if (!text) return "";
    return text.replace(/`/g, '\\`').replace(/\$/g, '\\$');
}

function extractSection(text, startMarker, nextMarkers) {
    const startIdx = text.indexOf(startMarker);
    if (startIdx === -1) return "";

    let contentStart = text.indexOf('\n', startIdx);
    if (contentStart === -1) contentStart = startIdx + startMarker.length;

    let endIdx = text.length;
    for (const marker of nextMarkers) {
        const pos = text.indexOf(marker, contentStart);
        if (pos !== -1 && pos < endIdx) {
            endIdx = pos;
        }
    }
    return text.substring(contentStart, endIdx).trim();
}

function parse() {
    if (!fs.existsSync(SOURCE_FILE)) {
        console.error(`Error: ${SOURCE_FILE} not found.`);
        return;
    }

    // Clear previous build to remove stale files (like old template-only pages)
    if (fs.existsSync(DOCS_DIR)) {
        fs.rmSync(DOCS_DIR, { recursive: true, force: true });
    }
    fs.mkdirSync(DOCS_DIR, { recursive: true });

    const content = fs.readFileSync(SOURCE_FILE, 'utf-8');
    const matches = [...content.matchAll(CATEGORY_PATTERN)];

    for (let i = 0; i < matches.length; i++) {
        const match = matches[i];
        const catId = match[1];
        const catName = match[2];
        const startPos = match.index + match[0].length;
        const endPos = (i + 1 < matches.length) ? matches[i + 1].index : content.length;

        const catContent = content.substring(startPos, endPos);
        const catNo = catId.replace('No.', '');
        const catDirName = `${catNo}-${slugify(catName)}`;
        const catPath = path.join(DOCS_DIR, catDirName);

        if (!fs.existsSync(catPath)) {
            fs.mkdirSync(catPath, { recursive: true });
        }

        // Create _category_.json for sidebar label
        const categoryMetadata = {
            label: `${catNo} ${catName}`,
            position: i + 1
        };
        fs.writeFileSync(path.join(catPath, '_category_.json'), JSON.stringify(categoryMetadata, null, 2));

        const blocks = catContent.split(DISEASE_PATTERN);

        let initialVisitTemplate = null;
        let followupVisitTemplate = null;
        const diseaseBlocks = [];

        // First pass: extract visit templates
        for (let block of blocks) {
            block = block.trim();
            if (!block) continue;

            if (block.includes('(初診)')) {
                initialVisitTemplate = {
                    sText: extractSection(block, '<S:', ['<O:', '<Ns:', '▶ Dr.Advice']),
                    oText: extractSection(block, '<O:', ['<Ns:', '▶ Dr.Advice']),
                    nsText: extractSection(block, '<Ns:', ['▶ Dr.Advice']),
                    advice: [...block.matchAll(/▶ Dr\.Advice (.*?)\n/g)].map(m => m[1].trim()).join('\n')
                };
            } else if (block.includes('(再診)')) {
                followupVisitTemplate = {
                    sText: extractSection(block, '<S:', ['<O:', '▶ Dr.Advice']),
                    oText: extractSection(block, '<O:', ['▶ Dr.Advice']),
                    advice: [...block.matchAll(/▶ Dr\.Advice (.*?)\n/g)].map(m => m[1].trim()).join('\n')
                };
            } else {
                diseaseBlocks.push(block);
            }
        }

        let diseaseCount = 0;
        for (let block of diseaseBlocks) {
            const titleMatch = block.match(/\[(.*?)\] (.*?)\n/);
            if (!titleMatch) continue;

            const icdCode = titleMatch[1];
            const diseaseName = titleMatch[2];
            diseaseCount++;

            // Local S/O extraction for blocks without (初診)/(再診) markers
            const localS = extractSection(block, '<S:', ['<O:', '<Ns:', '<A:', '<P:', '▶ Dr.Advice']);
            const localO = extractSection(block, '<O:', ['<Ns:', '<A:', '<P:', '▶ Dr.Advice']);
            const localNs = extractSection(block, '<Ns:', ['<A:', '<P:', '▶ Dr.Advice']);
            const localAdvice = [...block.matchAll(/▶ Dr\.Advice (.*?)(?:\n|$)/g)].map(m => m[1].trim()).join('\n');

            const assessment = extractSection(block, '<A:', ['<P:', '▶ Dr.Advice']);
            const plan = extractSection(block, '<P:', ['▶ Dr.Advice']);

            // Fallback logic
            const initialS = localS || initialVisitTemplate?.sText || "";
            const initialO = localO || initialVisitTemplate?.oText || "";
            const initialNs = localNs || initialVisitTemplate?.nsText || "";
            const initialAdv = localAdvice || initialVisitTemplate?.advice || "";

            const safeName = slugify(`${icdCode}-${diseaseName}`);
            const filePath = path.join(catPath, `${safeName}.mdx`);

            const mdxContent = `---
title: "[${icdCode}] ${diseaseName}"
sidebar_position: ${diseaseCount}
---

import MedicalEntry from '@site/src/components/MedicalEntry';

<MedicalEntry
  diseaseName="${escapeJS(diseaseName)}"
  icdCode="${escapeJS(icdCode)}"
  initialVisit={{
    sText: \`${escapeJS(initialS)}\`,
    oText: \`${escapeJS(initialO)}\`,
    nsText: \`${escapeJS(initialNs)}\`,
    advice: \`${escapeJS(initialAdv)}\`
  }}
  followupVisit={{
    sText: \`${escapeJS(followupVisitTemplate?.sText || "")}\`,
    oText: \`${escapeJS(followupVisitTemplate?.oText || "")}\`,
    advice: \`${escapeJS(followupVisitTemplate?.advice || "")}\`
  }}
  assessment={\`${escapeJS(assessment)}\`}
  plan={\`${escapeJS(plan)}\`}
  drAdvice={\`${escapeJS(localAdvice)}\`}
>
  :::info Wiki編集エリア
  この右側のエリアは、最新のガイドラインや病態生理、エビデンスを追記するためのスペースです。
  ファイルを直接編集して追記・修正を行ってください。
  :::
  
  ### 🔗 関連リンク・文献
  - [国内ガイドライン検索](https://www.google.com/search?q=${encodeURIComponent(diseaseName)}+ガイドライン)
</MedicalEntry>
`;
            fs.writeFileSync(filePath, mdxContent);
        }
    }
    console.log("Parsing completed with improved sidebar labels and fallback S/O logic.");
}

parse();
